### ISSUE TITLE
Differentiating between inline link targets and link targets from references?

### ISSUE THREAD

**Issue by afcowie** [jgm/pandoc#5390]
It appears the current AST can't differentiate between links that are defined with reference and links that are defined inline, the Markdown reader appears to substitute the URLs and titles in inline.

ie, pattern matching on 

```haskell
    ...
     Link attr inlines target -> handleLinks attr inlines target
    ...
```
can't tell me whether the source Markdown defined 

```markdown
This [link][1] is really something you should click on!

[1]: https://clickbait.federal.gov/
```
or

```markdown
If you actually put the [link](https://clickbait.federal.gov/) inline people
are less likely to enjoy reading your document.
```

Incidentally, in the Markdown writer it was really rather surprising to not have

```markdown
![Sexy person of appropriate gender](/images/photo.jpg)

Lorem ipsum yada yada...
```

but instead have

```markdown
![Sexy person of appropriate gender]

Lorem ipsum yada yada...

[Sexy person of appropriate gender]: /images/photo.jpg
```

when I did `pandoc -f markdown -t markdown ...`

I'm writing a tool to (re)format Markdown documents, ie roundtripping. I am using **pandoc**'s Text.Pandoc.Readers.Markdown reader on the way in, and have (with great help from your helper functions) written what is effectively an alternative writer for serializing out. Everything Was Fine™ until I got to the references vs inlines case for links (and images).

Based on your comments in a few other issues around I'm going to guess that you're not in a hurry to change the types in Text.Pandoc.Definition to support this extra information, but maybe it's there and I'm just missing something obvious.

If you _would_ support a change, I would be happy to write the patch per your direction. Before it got to the point of needing to change/add a constructor in `Inline` (say), perhaps we could bury the reference origin information in the atttributes mess or hide it in the target `(String,String), or...

May regards for such a fabulous project. It's been a privilege coding against Text.Pandoc and studying your code. Your implementations always seem more elegant than mine.

AfC

**Comment by jgm** [comment]
That's correct, this information is not recorded.  Neither are a lot of other details that an author might want to preserve in a "reformatting" step, e.g. whether a compact or shortcut style reference link was used, what the reference link key was, whether setext or atx headers are used, what bullet list marker is used, list item indentation, whether an indented code block or a fenced code block with no attribute was used, whether an autolink like `<http://example.com>` or an explicit link was used, what characters were used for a horizontal rule, whether spaces or tabs were used for indentation, indentation of raw HTML, whether a fenced div or HTML-style "native div" was used for a Div element, whether YAML metadata was all in one block, or in several, and how it was positioned in the document, etc.

I just mention these things because the same reasons you have for wanting to keep track of reference links may apply to them.  Pandoc wasn't really designed to be used as a code reformatter or linter; I think for these purposes one would want a quite different parser design.  (commonmark-hs, which I'm working on, may be more suitable.)

**Comment by afcowie** [comment]
That all makes sense. I hadn't run into most of those because I am normalizing the input documents to a locally canonical form (also known as "what I think looks right, take that the rest of you"), so _e.g._ preserving input bullet types was a non-issue.

Hm.

Darn it. I was so close!

For now I might cop out by writing a pre-processor that lifts out all the reference declarations and then then when serializing out restores them. ish.

I will follow up your suggestion about **commonmark-hs**, though last I looked it didn't have support for most of the juicy Pandoc extensions (which is because it's _commonmark_, yo) [the one I noticed off hand was the commonmark spec lacking definition lists]. I'm certainly not in a rush to write a new parser. Perhaps I should consider extending **commonmark-hs** to include things I discover I care about? [the first task of which would be figuring out what that list even is]. If you suggest going down that road I'll open conversations there or on a list as appropriate.

Thanks for your reply above.
